pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,180 @@
+package routemetrics
+
+import (
+	"context"
+	"fmt"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+)
+
+const (
+	controllerName = "route_metrics_controller"
+)
+
+var (
+	log = logf.Logger.WithName(controllerName)
+)
+
+// New creates the route metrics controller. This is the controller
+// that handles all the logic for gathering and exporting
+// metrics related to route resources.
+func New(mgr manager.Manager, namespace string) (controller.Controller, error) {
+	// Create a copy of the cluster, but with a new cache to watch on Route objects from every namespace.
+	newCluster, err := cluster.New(mgr.GetConfig(), func(o *cluster.Options) {
+		o.Scheme = mgr.GetScheme()
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Add the cluster to the manager so that the cluster is started along with the other runnables.
+	mgr.Add(newCluster)
+	reconciler := &reconciler{
+		client:    mgr.GetClient(),
+		Namespace: namespace,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	// add watch for changes in IngressController
+	if err := c.Watch(&source.Kind{Type: &operatorv1.IngressController{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	// add watch for changes in Route
+	if err := c.Watch(source.NewKindWithCache(&routev1.Route{}, newCluster.GetCache()),
+		handler.EnqueueRequestsFromMapFunc(reconciler.routeToIngressController)); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// routeToIngressController .
```

Comment:
Please add a bit more to the comment to explain this function's purpose.  If you're already watching for changes in IngressController, then why do you need to watch Routes too?  Conversely, if you watch Routes, why do you need to watch for changes in IngressController?

Reply:
I missed adding the comments for the function. I'll add them.

Tracking changes for both the kind of objects are required. When a Ingress Controller is created, there may not be any Routes which will be assigned to it. Thus, we need to watch for Ingress controller events so that we can initialize the corresponding labels of the metric. Whenever a Route is assigned (unassigned) to an Ingress Controller, the reconcile loop needs to run. Thus, we need to watch for Route events as well.

Reply:
Okay, thanks.  I discussed this with the team, and I'm going to update some of my comments.  We think it will be better if you follow the design pattern used by metrics collection in the Canary controller, using a probe and polling like in https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/canary/controller.go#L191

Reply:
The Canary Controller probes for a single Route object named `canary` in the `openshift-ingress-canary` namespace. We would not be able to probe for particular routes as we would not know the name beforehand. A watch on the Route object would be required so that we can get events related to any Route object in any namespace. 

Additionally, we would need to reconcile for the Ingress Controllers related to the Route object for which a CRUD event occurred, so that we can properly update the `RoutesPerShard` metric for each of them.